create_role/modules/deductive_role/main.tf

```diff
@@ -0,0 +1,678 @@
+#    - Permissions: GetSecretValue, DescribeSecret, AssumeRole (EC2Role)
+#    - Purpose: Allows Kubernetes pods to read secrets from AWS Secrets Manager
+#
+# 5. SecretsWriterReaderRole - Role for reading and writing secrets to AWS Secrets Manager
+#    - Permissions: GetSecretValue, DescribeSecret, CreateSecret, PutSecretValue, UpdateSecret
+#    - Purpose: Allows Kubernetes pods to read and write secrets to AWS Secrets Manager
+#
+# Each role has specific policies attached that grant the minimum necessary permissions
+# for DeductiveAI to operate effectively while maintaining security best practices.
+#
+# The trust relationships are configured as follows:
+# - DeductiveAssumeRole: Trusted by DeductiveAI AWS account
+# - EKSClusterRole: Trusted by EKS service
+# - EC2Role: Trusted by EC2 service and SecretsReaderRole
+# - SecretsReaderRole: Initially trusted by EKS service, later updated for OIDC
+# - SecretsWriterReaderRole: Initially trusted by EKS service, later updated for OIDC
+
+provider "aws" {
+  region  = var.region
+  profile = var.aws_profile
+}
+
+data "aws_caller_identity" "current" {}
+
+# Define the assume role policy document
+data "aws_iam_policy_document" "assume_role_policy" {
+  statement {
+    effect = "Allow"
+    principals {
+      type        = "AWS"
+      identifiers = ["arn:aws:iam::${local.deductive_aws_account_id}:root"]
+    }
+    actions = ["sts:AssumeRole"]
+  }
+}
+
+# Define the main policy document for Deductive operations
+data "aws_iam_policy_document" "deductive_policy" {
+  # Read-only permissions for various AWS services
+  statement {
+    effect = "Allow"
+    actions = [
+      # EC2 Read-Only Actions
+      "ec2:Describe*",
+      "ec2:Get*",
+      "ec2:List*",
+      # ELB Read-Only Actions
+      "elasticloadbalancing:Describe*",
+      # EKS Read-Only Actions
+      "eks:Describe*",
+      "eks:List*",
+      # IAM Read-Only Actions
+      "iam:GetRole",
+      "iam:GetInstanceProfile",
+      "iam:GetPolicy",
+      "iam:GetPolicyVersion",
+      "iam:GetRolePolicy",
+      "iam:ListAttachedRolePolicies",
+      "iam:ListEntitiesForPolicy",
+      "iam:ListInstanceProfilesForRole",
+      "iam:ListRolePolicies",
+      "iam:ListPolicyVersions",
+      "iam:CreatePolicyVersion",
+      # For nodegroup creation
+      "iam:CreateServiceLinkedRole",
+      # ACM Read-Only Actions
+      "acm:Describe*",
+      "acm:List*"
+    ]
+    resources = ["*"]
+  }
+
+  # Create Deductive SSL certificate for SSO
+  statement {
+    effect = "Allow"
+    actions = [
+      "acm:RequestCertificate"
+    ]
+    resources = ["arn:aws:acm:*:${data.aws_caller_identity.current.account_id}:certificate/*"]
+    condition {
+      test     = "StringEquals"
+      variable = "aws:RequestTag/creator"
+      values   = ["deductive-ai"]
+    }
+  }
+
+  # Create permission to run instances
+  statement {
+    effect = "Allow"
+    actions = [
+      "ec2:RebootInstances",
+      "ec2:RunInstances",
```

Comment:
If we use launchtemplate, we may need RunInstances to apply on a broader resource

Reply:
as discuss, let's have a follow up on this with state management